CRM/Core/Error.php

```diff
@@ -1099,6 +1099,10 @@ public static function isAPIError($error, $type = CRM_Core_Error::FATAL_ERROR) {
    *   optional description of old method (if not the calling method). eg. CRM_MyClass::myOldMethodToGetTheOptions()
    */
   public static function deprecatedFunctionWarning(string $newMethod, ?string $oldMethod = NULL): void {
+    if (!(error_reporting() & E_USER_DEPRECATED)) {
+      // Logging deprecations should be opt-in, it's useless to non-developers.
+      return;
+    }
```

Comment:
I did some tests and I just want to clarify log vs onscreen message. The function logs unconditionally to civi log, but only displays on screen if error_reporting already includes E_USER_DEPRECATED (the trigger_error, which is handled by e.g. drupal).

So I partly agree with this but I think it's in the wrong spot. It should only be guarding the call to Civi::log down below on line 1115 (and 1131). The trigger_error should always happen because it's already filtered to E_USER_DEPRECATED so only has any effect if php.ini error_reporting includes E_USER_DEPRECATED, which it wouldn't on a typical production site. If it does and you don't have access to php.ini then error_reporting can be set in drupal settings.php (civicrm.settings.php is too late because it's drupal that's displaying the message on screen).

@eileenmcnaughton are you saying WMF wants these in the civi log even when E_USER_DEPRECATED is not included in php.ini?

Reply:
@demeritcowboy yep - the E_USER_DEPRECATED controls whether they wind up in the syslog but I wouldn't want these messages to be any less prominent because when we see them we take action, or plan to & these messages highlight issues we don't always hit on our local because we don't have exactly the same data or do exactly the same things as our users to.

We also benefit from the fact that we often have deprecation related issues & fixes  logged against extensions that we manage by other organizations, some of which are picking those errors up from their logs

Reply:
Ok. I see both sides. @agileware-fj while it might not be ideal would this work for you in an extension:

```php
myext_civicrm_container($container) {
  $container->setDefinition('psr_log', new \Symfony\Component\DependencyInjection\Definition('\Civi\MyExt\Logger', []))->setPublic(TRUE);
}

class Civi\MyExt\Logger extends CRM_Core_Error_Log {
  public function log($level, $message, array $context = []): void {
    // copy all the stuff from CRM_Core_Error_Log::log, but exclude logging anything that has
    // `'civi.tag' => 'deprecated'` in the $context array if error_reporting() doesn't have E_USER_DEPRECATED
  }
```

@eileenmcnaughton I realize this conflicts with your second point but that is what seems to be the conflict here.

Reply:
This is a solution that works for us, and the PR is here because I think for the community as a whole, it's better not to massively increase the size of log files (literally 50× on one system we were just looking at this morning) because ... mostly, permissions have been defined correctly.

I know when I'm trying to solve a problem that probably has most information required in the CiviCRM log, and it's full of deprecation messages, I actually despair and become repulsed from continuing to look in the log file for the actual relevant information, not necessarily because the deprecation notices are there, but because they are most of the log file contents. I expect there's other people working on CiviCRM sites out there who experience the same anguish.

If visibility is so important, perhaps it would be worth storing these deprecations in cache and generate check notices for them?